HW2/P2/parallel_vector.pyx

```diff
@@ -79,12 +79,20 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
+
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads = 4):
+
                if counts[src[idx]] > 0:
+                   acquire(&(locks[dest[idx]]))
```

Comment:
acquiring the lock here may be too late, and counts[src[idx]] may have been invalidated since checking the condition (e.g., it is no longer > 0)